File: server/models/Etudiant.js
const mongoose = require('mongoose');
const bcrypt = require('bcryptjs');

const etudiantSchema = new mongoose.Schema({
  matricule: {
    type: String,
    required: true,
    unique: true
  },
  nom: {
    type: String,
    required: true
  },
  prenom: {
    type: String,
    required: true
  },
  email: {
    type: String,
    required: true,
    unique: true
  },
  telephone: {
    type: String,
    required: true
  },
  adresse: {
    type: String
  },
  mot_de_passe: {
    type: String,
    required: true
  },
  programme_id: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Programme',
    required: true
  },
  groupe: {
    type: Number,
    required: true
  },
  qr_code: {
    type: String,
    unique: true
  },
  chef_classe_validateur: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'ChefDeClasse'
  },
  date_validation_compte: {
    type: Date
  },
  pourcentage_paiement: {
    type: Number,
    default: 0,
    min: 0,
    max: 100
  },
  pourcentage_paiement_seuil: {
    type: Number,
    default: 75,
    min: 0,
    max: 100
  },
  derniere_entree_fac: {
    type: Date
  }
}, {
  timestamps: true
});

// Hash password
etudiantSchema.pre('save', async function(next) {
  if (!this.isModified('mot_de_passe')) return next();
  this.mot_de_passe = await bcrypt.hash(this.mot_de_passe, 10);
  next();
});

// Générer QR code unique
etudiantSchema.pre('save', async function(next) {
  if (!this.qr_code) {
    this.qr_code = `STU${Date.now()}${Math.random().toString(36).substr(2, 5)}`.toUpperCase();
  }
  next();
});

module.exports = mongoose.model('Etudiant', etudiantSchema);

File: eduflex_mobile/lib/models/etudiant.dart
class Etudiant {
  final String id;
  final String matricule;
  final String nom;
  final String prenom;
  final String email;
  final String telephone;
  final String qrCode;
  final String statutCompte;
  final Programme? programme;
  final int groupe;
  final DateTime? derniereEntreeFac;

  Etudiant({
    required this.id,
    required this.matricule,
    required this.nom,
    required this.prenom,
    required this.email,
    required this.telephone,
    required this.qrCode,
    required this.statutCompte,
    this.programme,
    required this.groupe,
    this.derniereEntreeFac,
  });

  factory Etudiant.fromJson(Map<String, dynamic> json) {
    return Etudiant(
      id: json['_id'] ?? json['id'] ?? '',
      matricule: json['matricule'] ?? '',
      nom: json['nom'] ?? '',
      prenom: json['prenom'] ?? '',
      email: json['email'] ?? '',
      telephone: json['telephone'] ?? '',
      qrCode: json['qr_code'] ?? '',
      statutCompte: json['statut_compte'] ?? 'en_attente',
      programme: json['programme_id'] != null
          ? Programme.fromJson(json['programme_id'])
          : null,
      groupe: json['groupe'] ?? 1,
      derniereEntreeFac: json['derniere_entree_fac'] != null
          ? DateTime.parse(json['derniere_entree_fac'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'matricule': matricule,
      'nom': nom,
      'prenom': prenom,
      'email': email,
      'telephone': telephone,
      'qr_code': qrCode,
      'statut_compte': statutCompte,
      'programme_id': programme?.toJson(),
      'groupe': groupe,
      'derniere_entree_fac': derniereEntreeFac?.toIso8601String(),
    };
  }
}

class Programme {
  final String id;
  final String nom;
  final String licence;
  final int semestre;

  Programme({
    required this.id,
    required this.nom,
    required this.licence,
    required this.semestre,
  });

  factory Programme.fromJson(Map<String, dynamic> json) {
    return Programme(
      id: json['_id'] ?? json['id'] ?? '',
      nom: json['nom'] ?? '',
      licence: json['licence'] ?? '',
      semestre: json['semestre'] ?? 1,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'nom': nom,
      'licence': licence,
      'semestre': semestre,
    };
  }
}
